linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Linker.scala

```diff
@@ -12,19 +12,197 @@
+        if (report.publicModules.size != 1) {
+          throw new LinkingException(
+              "Linking did not return exactly one public module, but the legacy " +
+              "`link` method was called. Call the overload taking an " +
+              s"OutputDirectory instead. Full report:\n$report")
+        }
+
+        report.publicModules.head
+      }
+
+      val expectedFiles = Set(module.jsFileName) ++ module.sourceMapName
+      val foundFiles = outDir.content.keySet
+
+      if (foundFiles != expectedFiles) {
+        if (expectedFiles.subsetOf(foundFiles)) {
+          throw new LinkingException(
+              "Linking produced more than a single JS file (and source map) but " +
+              "the legacy `link` method was called. Call the overload taking " +
+              "an OutputDirectory instead. " +
+              s"Expected files:\n$expectedFiles\nProduced files:\n$foundFiles")
+        } else {
+          throw new LinkingException(
+              "Linking did not produce the files mentioned in the report. " +
+              "This is a bug in the linker." +
+              s"Expected files:\n$expectedFiles\nProduced files:\n$foundFiles")
+        }
+      }
+
+      val hasSourceMap =
+        module.sourceMapName.isDefined && output.sourceMap.isDefined
+
+      import StandardCharsets.UTF_8
+
+      val jsFileWrite = {
+        val content = new String(outDir.content(module.jsFileName), UTF_8)
+        val patched = Linker.patchJSFileContent(content,
+            output.sourceMapURI.filter(_ => hasSourceMap))
+
+        OutputFileImpl.fromOutputFile(output.jsFile)
+          .writeFull(ByteBuffer.wrap(patched.getBytes(UTF_8)))
+      }
+
+      val sourceMapWrite = for {
+        sourceMapName <- module.sourceMapName
+        sourceMapFile <- output.sourceMap
+      } yield {
+        val content = new String(outDir.content(sourceMapName), UTF_8)
+        val patched = Linker.patchSourceMapContent(content, output.jsFileURI)
+
+        OutputFileImpl.fromOutputFile(sourceMapFile)
+          .writeFull(ByteBuffer.wrap(patched.getBytes(UTF_8)))
+      }
+
+      Future.sequence(List(jsFileWrite) ++ sourceMapWrite).map(_ => ())
+    }
+  }
+}
+
+private object Linker {
+  private val sourceMapRe = """(?m)^//# sourceMappingURL=.*$""".r
+
+  /* It is somewhat acceptable to parse the JSON field "file" with a Regex
+   * because:
+   *
+   * - Source maps do not contain nested objects.
+   * - The file URI should not contain '"', because URI.toASCIIString (which is
+   *   used by the StandardLinker) never returns them.
+   *
+   * So as a legacy mechanism, this is OK-ish. It keeps us from having to build
+   * the infrastructure to parse JSON cross platform.
+   */
+  private val fileFieldRe = """(?m)([,{])\s*"file"\s*:\s*".*"\s*([,}])""".r
+
+  /** Patches the JS file content to have the provided source map link (or none) */
+  private def patchJSFileContent(content: String,
+      sourceMapURI: Option[URI]): String = {
```

Comment:
Could you elaborate a bit on why those patches are necessary, and what they do exactly?